<a href="https://colab.research.google.com/github/PRihana/Machine-learning/blob/master/23_03_2020(ML)pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np
import pandas as pd

In [3]:
# numpy version check
print(np.__version__)
# Pandas version
print(pd.__version__)


1.18.2
0.25.3


In [0]:
# Read the data
data = pd.read_csv("/FL_insurance_sample.csv.txt")

In [7]:
# Checking output type
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [8]:
# To check the shape 
print(data.shape)
# to get the number of rows or instances or records or examples
print(data.shape[0])
# to get the number of columns or features or attributes
print(data.shape[1])

(36634, 18)
36634
18


In [9]:
# To get the column names
print(data.columns)
# To get the row names
print(data.index)

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')
RangeIndex(start=0, stop=36634, step=1)


In [10]:
# Datatype of each column
print(data.dtypes)
print("-----"*15)
# what is output type for data.dtypes
print(type(data.dtypes))
print("-----"*15)
# Access the first column dtype
print(data.dtypes[0])
print("-----"*15)
# Access the last column dtype
print(data.dtypes[-1])

policyID                int64
statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible      int64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity       int64
dtype: object
---------------------------------------------------------------------------
<class 'pandas.core.series.Series'>
---------------------------------------------------------------------------
int64
---------------------------------------------------------------------------
int64


In [12]:
# Column count for each datatype
# How many columns are there with int64
# How many columns are there with float64
# How many columns are there with object
data.dtypes.value_counts()
# Note We are using series method value_counts

float64    11
object      4
int64       3
dtype: int64

**observe the data**

In [13]:
# display top 5 rows
data.head()

policyID statecode       county  ...         line  construction  point_granularity
0    119736        FL  CLAY COUNTY  ...  Residential       Masonry                  1
1    448094        FL  CLAY COUNTY  ...  Residential       Masonry                  3
2    206893        FL  CLAY COUNTY  ...  Residential          Wood                  1
3    333743        FL  CLAY COUNTY  ...  Residential          Wood                  3
4    172534        FL  CLAY COUNTY  ...  Residential          Wood                  1

[5 rows x 18 columns]

In [14]:
# display top 10 rows
data.head(10)

policyID statecode  ...         construction  point_granularity
0    119736        FL  ...              Masonry                  1
1    448094        FL  ...              Masonry                  3
2    206893        FL  ...                 Wood                  1
3    333743        FL  ...                 Wood                  3
4    172534        FL  ...                 Wood                  1
5    785275        FL  ...              Masonry                  3
6    995932        FL  ...  Reinforced Concrete                  1
7    223488        FL  ...                 Wood                  1
8    433512        FL  ...                 Wood                  1
9    142071        FL  ...              Masonry                  1

[10 rows x 18 columns]

In [15]:
# display last 5 rows
data.tail()

policyID statecode  ... construction  point_granularity
36629    404309        FL  ...         Wood                  3
36630    465875        FL  ...      Masonry                  1
36631    791209        FL  ...         Wood                  4
36632    322627        FL  ...      Masonry                  3
36633    398149        FL  ...      Masonry                  1

[5 rows x 18 columns]

In [16]:
# display last 3 rows
data.tail(3)

policyID statecode  ... construction  point_granularity
36631    791209        FL  ...         Wood                  4
36632    322627        FL  ...      Masonry                  3
36633    398149        FL  ...      Masonry                  1

[3 rows x 18 columns]

In [17]:
# select a random row
data.sample()

policyID statecode  ... construction  point_granularity
36390    839109        FL  ...         Wood                  1

[1 rows x 18 columns]

In [18]:
# select 5 rows randomly
data.sample(5)

policyID statecode  ... construction  point_granularity
30284    439552        FL  ...         Wood                  3
18705    352979        FL  ...      Masonry                  1
11385    178194        FL  ...      Masonry                  1
13292    160039        FL  ...      Masonry                  1
32153    205605        FL  ...         Wood                  4

[5 rows x 18 columns]

- data.head()
- data.tail()
-data.samle()

the output type of all the above methods are dataframe

In [19]:
data.head(1)

policyID statecode       county  ...         line  construction  point_granularity
0    119736        FL  CLAY COUNTY  ...  Residential       Masonry                  1

[1 rows x 18 columns]

In [22]:
data.tail(1)

policyID statecode  ... construction  point_granularity
36633    398149        FL  ...      Masonry                  1

[1 rows x 18 columns]

- data.head(1)
- data.samle(1)

the output type of all the above methods are also dataframe which consists a single row

**Info about the dataframe**

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36634 entries, 0 to 36633
Data columns (total 18 columns):
policyID              36634 non-null int64
statecode             36634 non-null object
county                36634 non-null object
eq_site_limit         36634 non-null float64
hu_site_limit         36634 non-null float64
fl_site_limit         36634 non-null float64
fr_site_limit         36634 non-null float64
tiv_2011              36634 non-null float64
tiv_2012              36634 non-null float64
eq_site_deductible    36634 non-null float64
hu_site_deductible    36634 non-null float64
fl_site_deductible    36634 non-null float64
fr_site_deductible    36634 non-null int64
point_latitude        36634 non-null float64
point_longitude       36634 non-null float64
line                  36634 non-null object
construction          36634 non-null object
point_granularity     36634 non-null int64
dtypes: float64(11), int64(3), object(4)
memory usage: 5.0+ MB


In [24]:
type(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36634 entries, 0 to 36633
Data columns (total 18 columns):
policyID              36634 non-null int64
statecode             36634 non-null object
county                36634 non-null object
eq_site_limit         36634 non-null float64
hu_site_limit         36634 non-null float64
fl_site_limit         36634 non-null float64
fr_site_limit         36634 non-null float64
tiv_2011              36634 non-null float64
tiv_2012              36634 non-null float64
eq_site_deductible    36634 non-null float64
hu_site_deductible    36634 non-null float64
fl_site_deductible    36634 non-null float64
fr_site_deductible    36634 non-null int64
point_latitude        36634 non-null float64
point_longitude       36634 non-null float64
line                  36634 non-null object
construction          36634 non-null object
point_granularity     36634 non-null int64
dtypes: float64(11), int64(3), object(4)
memory usage: 5.0+ MB


NoneType

**Accesing a single column**

In [25]:
# To access the column policyID
data.policyID

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
36629    404309
36630    465875
36631    791209
36632    322627
36633    398149
Name: policyID, Length: 36634, dtype: int64

In [26]:
# What is the ouput type
print(type(data.policyID))

<class 'pandas.core.series.Series'>


In [27]:
# To access the column policyID
data['policyID']

0        119736
1        448094
2        206893
3        333743
4        172534
          ...  
36629    404309
36630    465875
36631    791209
36632    322627
36633    398149
Name: policyID, Length: 36634, dtype: int64

In [28]:
# What is the ouput type
print(type(data['policyID']))

<class 'pandas.core.series.Series'>


**specific column Datatype**

In [29]:
data.statecode

0        FL
1        FL
2        FL
3        FL
4        FL
         ..
36629    FL
36630    FL
36631    FL
36632    FL
36633    FL
Name: statecode, Length: 36634, dtype: object

In [30]:
# Specific column datatype
data.statecode.dtype

dtype('O')

In [31]:
data.statecode.dtypes

dtype('O')

In [32]:
data.dtype

AttributeError: ignored

difference between dtype and dtypes

- datatafrme object doesn't have dtype

- Numerical columns - int or float data

  - mean,median,mode,sum,quartiles,min,max

- Categorical columns - Object type(string data)

- Bollean data - True or False

In [33]:
data.point_granularity

0        1
1        3
2        1
3        3
4        1
        ..
36629    3
36630    1
36631    4
36632    3
36633    1
Name: point_granularity, Length: 36634, dtype: int64

In [34]:
data.point_granularity.dtype

dtype('int64')

In [35]:
# mean value of a column
data.point_granularity.mean()

1.6409073538243162

In [36]:
# mean value of a column
data.point_granularity.sum()/data.point_granularity.shape[0]

1.6409073538243162

In [37]:
# minimum value of a column
print(data.point_granularity.min())
# maximum value of a column
print(data.point_granularity.max())

1
7


In [38]:
data.point_granularity.value_counts()

1    26569
3     6724
4     3337
7        2
5        2
Name: point_granularity, dtype: int64

In [39]:
data.point_granularity.mode()
# output type is a series

0    1
dtype: int64

In [40]:
data.point_granularity.mode()[0]

1

In [41]:
data.point_granularity.median()

1.0

In [42]:
# Range of values in a column
(data.point_granularity.min(),data.point_granularity.max())

(1, 7)

In [43]:
data.point_granularity.count()

36634

In [44]:
data.county.dtype

dtype('O')

In [45]:
# How many non null values are there
data.county.count()
# Output type is an integer

36634

In [46]:
# For each uniwue value what is the count
data.county.value_counts()
# output type is a series
# In that series the labels are unique values
# the value for that label is how many times it occured in that column

MIAMI DADE COUNTY    4315
BROWARD COUNTY       3193
PALM BEACH COUNTY    2791
DUVAL COUNTY         1894
ORANGE COUNTY        1811
                     ... 
GLADES COUNTY          22
UNION COUNTY           15
OSCEOLA COUNTY          1
Orlando                 1
North Fort Myers        1
Name: county, Length: 67, dtype: int64

In [47]:
# How many unique values are there
print(data.county.nunique())
print(data.county.value_counts().shape[0])

67
67


In [48]:
# What are the uinique values
print(data.county.unique())
print(data.county.value_counts().index)

['CLAY COUNTY' 'SUWANNEE COUNTY' 'NASSAU COUNTY' 'COLUMBIA COUNTY'
 'ST  JOHNS COUNTY' 'BAKER COUNTY' 'BRADFORD COUNTY' 'HAMILTON COUNTY'
 'UNION COUNTY' 'MADISON COUNTY' 'LAFAYETTE COUNTY' 'FLAGLER COUNTY'
 'DUVAL COUNTY' 'LAKE COUNTY' 'VOLUSIA COUNTY' 'PUTNAM COUNTY'
 'MARION COUNTY' 'SUMTER COUNTY' 'LEON COUNTY' 'FRANKLIN COUNTY'
 'LIBERTY COUNTY' 'GADSDEN COUNTY' 'WAKULLA COUNTY' 'JEFFERSON COUNTY'
 'TAYLOR COUNTY' 'BAY COUNTY' 'WALTON COUNTY' 'JACKSON COUNTY'
 'CALHOUN COUNTY' 'HOLMES COUNTY' 'WASHINGTON COUNTY' 'GULF COUNTY'
 'ESCAMBIA COUNTY' 'SANTA ROSA COUNTY' 'OKALOOSA COUNTY' 'ALACHUA COUNTY'
 'GILCHRIST COUNTY' 'LEVY COUNTY' 'DIXIE COUNTY' 'SEMINOLE COUNTY'
 'ORANGE COUNTY' 'BREVARD COUNTY' 'INDIAN RIVER COUNTY'
 'MIAMI DADE COUNTY' 'BROWARD COUNTY' 'MONROE COUNTY' 'PALM BEACH COUNTY'
 'MARTIN COUNTY' 'HENDRY COUNTY' 'PASCO COUNTY' 'GLADES COUNTY'
 'HILLSBOROUGH COUNTY' 'HERNANDO COUNTY' 'PINELLAS COUNTY' 'POLK COUNTY'
 'North Fort Myers' 'Orlando' 'HIGHLANDS COUNTY' 'HARDE

In [49]:
# To get the description for the numerical columns
data.describe()

policyID  eq_site_limit  ...  point_longitude  point_granularity
count   36634.000000   3.663400e+04  ...     36634.000000       36634.000000
mean   548661.937189   7.314780e+05  ...       -81.903565           1.640907
std    259912.926494   1.752300e+07  ...         1.809066           1.070561
min    100074.000000   0.000000e+00  ...       -87.447290           1.000000
25%    323647.750000   0.000000e+00  ...       -82.439331           1.000000
50%    548525.000000   0.000000e+00  ...       -81.585700           1.000000
75%    774252.000000   0.000000e+00  ...       -80.372910           3.000000
max    999971.000000   2.160000e+09  ...       -80.033257           7.000000

[8 rows x 14 columns]

In [50]:
# To get the list of numerical columns
data.describe().columns

Index(['policyID', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit',
       'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible',
       'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible',
       'point_latitude', 'point_longitude', 'point_granularity'],
      dtype='object')

In [51]:
# To get the description for the categorical columns
data.describe(include=object)

statecode             county         line construction
count      36634              36634        36634        36634
unique         1                 67            2            5
top           FL  MIAMI DADE COUNTY  Residential         Wood
freq       36634               4315        30838        21581

In [52]:
data.describe(include=np.int64)

policyID  fr_site_deductible  point_granularity
count   36634.000000        36634.000000       36634.000000
mean   548661.937189           26.483594           1.640907
std    259912.926494         4702.740932           1.070561
min    100074.000000            0.000000           1.000000
25%    323647.750000            0.000000           1.000000
50%    548525.000000            0.000000           1.000000
75%    774252.000000            0.000000           3.000000
max    999971.000000       900000.000000           7.000000

In [53]:
data.describe(include=np.float64)

eq_site_limit  hu_site_limit  ...  point_latitude  point_longitude
count   3.663400e+04   3.663400e+04  ...    36634.000000     36634.000000
mean    7.314780e+05   2.074348e+06  ...       28.087477       -81.903565
std     1.752300e+07   1.964150e+07  ...        1.647734         1.809066
min     0.000000e+00   0.000000e+00  ...       24.547514       -87.447290
25%     0.000000e+00   4.248106e+04  ...       26.471996       -82.439331
50%     0.000000e+00   1.926911e+05  ...       28.057077       -81.585700
75%     0.000000e+00   9.163381e+05  ...       29.623005       -80.372910
max     2.160000e+09   2.160000e+09  ...       30.989820       -80.033257

[8 rows x 11 columns]

In [54]:
data.describe(include='all')

policyID statecode  ... construction  point_granularity
count    36634.000000     36634  ...        36634       36634.000000
unique            NaN         1  ...            5                NaN
top               NaN        FL  ...         Wood                NaN
freq              NaN     36634  ...        21581                NaN
mean    548661.937189       NaN  ...          NaN           1.640907
std     259912.926494       NaN  ...          NaN           1.070561
min     100074.000000       NaN  ...          NaN           1.000000
25%     323647.750000       NaN  ...          NaN           1.000000
50%     548525.000000       NaN  ...          NaN           1.000000
75%     774252.000000       NaN  ...          NaN           3.000000
max     999971.000000       NaN  ...          NaN           7.000000

[11 rows x 18 columns]

In [55]:
data.columns

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')

Slicing, subsetting, Boolean Indexing

- postion based (iloc)
  -    rows --> 0 to 36633
  -    columns --> 0 to 17
- label based indexing(loc)
  -    rows --> 0 to 36633
  -    columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']

In [56]:
data.iloc[1,:] # row with index 1 and all the columns

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [57]:
data.loc[1,:]

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [58]:
data.iloc[1,1] # row with index 1and column with index 1

'FL'

In [59]:
data.loc[1,"statecode"]

'FL'

In [60]:
data.iloc[1,1:3] # row with index 1 and columns with index 1,2
# iloc stop index is exclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [61]:
data.loc[1,"statecode":"county"] 
# loc - the stop label also inclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [62]:
data.iloc[1:4,1:3] 
# rows with indices 1,2,3
# columns with indices 1,2

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [63]:
data.loc[1:3,"statecode":"county"]

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [64]:
data.policyID.nunique()

36634

In [65]:
data.set_index("policyID")

statecode           county  ...  construction  point_granularity
policyID                             ...                                 
119736          FL      CLAY COUNTY  ...       Masonry                  1
448094          FL      CLAY COUNTY  ...       Masonry                  3
206893          FL      CLAY COUNTY  ...          Wood                  1
333743          FL      CLAY COUNTY  ...          Wood                  3
172534          FL      CLAY COUNTY  ...          Wood                  1
...            ...              ...  ...           ...                ...
404309          FL  PINELLAS COUNTY  ...          Wood                  3
465875          FL  PINELLAS COUNTY  ...       Masonry                  1
791209          FL  PINELLAS COUNTY  ...          Wood                  4
322627          FL  PINELLAS COUNTY  ...       Masonry                  3
398149          FL  PINELLAS COUNTY  ...       Masonry                  1

[36634 rows x 17 columns]

In [66]:
data

policyID statecode  ... construction  point_granularity
0        119736        FL  ...      Masonry                  1
1        448094        FL  ...      Masonry                  3
2        206893        FL  ...         Wood                  1
3        333743        FL  ...         Wood                  3
4        172534        FL  ...         Wood                  1
...         ...       ...  ...          ...                ...
36629    404309        FL  ...         Wood                  3
36630    465875        FL  ...      Masonry                  1
36631    791209        FL  ...         Wood                  4
36632    322627        FL  ...      Masonry                  3
36633    398149        FL  ...      Masonry                  1

[36634 rows x 18 columns]

In [67]:
df = data.set_index("policyID")
df.head()

statecode       county  ...  construction  point_granularity
policyID                         ...                                 
119736          FL  CLAY COUNTY  ...       Masonry                  1
448094          FL  CLAY COUNTY  ...       Masonry                  3
206893          FL  CLAY COUNTY  ...          Wood                  1
333743          FL  CLAY COUNTY  ...          Wood                  3
172534          FL  CLAY COUNTY  ...          Wood                  1

[5 rows x 17 columns]

In [68]:
df.index

Int64Index([119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
            433512, 142071,
            ...
            827844, 876385, 167630, 241496, 780814, 404309, 465875, 791209,
            322627, 398149],
           dtype='int64', name='policyID', length=36634)

- Slicing, subsetting, Boolean Indexing with df

- postion based (iloc)
  - rows --> 0 to 36633
  - columns --> 0 to 17
- label based indexing(loc)
  - rows --> [119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
  433512, 142071,...]
  - columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']

In [77]:
#row with index 1 and all the columns
data.iloc[1 ,: ]

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [78]:
# row with index 1and column with index 1
data.iloc[1 ,1 ]

'FL'

In [84]:
#row with index 1 and columns with index 1,2
data.iloc[1,1:3 ]

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [82]:
# rows with indices 1,2,3
# columns with indices 1,2
data.iloc[1:4 ,1:3 ]

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY